File: src/main/java/org/intellij/sdk/language/jass/JassFormattingModelBuilder.java
package org.intellij.sdk.language.jass;

import com.intellij.formatting.*;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import guru.xgm.jass.lang.Language_JASS;
import org.intellij.sdk.language.jass.psi.JassTypes;
import org.jetbrains.annotations.NotNull;

final class JassFormattingModelBuilder implements FormattingModelBuilder {

    private static SpacingBuilder createSpaceBuilder(CodeStyleSettings settings) {
        return new SpacingBuilder(settings, Language_JASS.INSTANCE)
                .around(JassTypes.SEPARATOR)
                .spaceIf(settings.getCommonSettings(Language_JASS.INSTANCE.getID()).SPACE_AROUND_ASSIGNMENT_OPERATORS)
                .before(JassTypes.PROPERTY)
                .none();
    }

    @Override
    public @NotNull FormattingModel createModel(@NotNull FormattingContext formattingContext) {
        final CodeStyleSettings codeStyleSettings = formattingContext.getCodeStyleSettings();
        return FormattingModelProvider
                .createFormattingModelForPsiFile(formattingContext.getContainingFile(),
                        new JassBlock(formattingContext.getNode(),
                                Wrap.createWrap(WrapType.NONE, false),
                                Alignment.createAlignment(),
                                createSpaceBuilder(codeStyleSettings)),
                        codeStyleSettings);
    }

}

File: src/main/kotlin/raft/war/language/jass/formatting/JassFormattingModelBuilder.kt
package raft.war.language.jass.formatting

import com.intellij.formatting.FormattingContext
import com.intellij.formatting.FormattingModel
import com.intellij.formatting.FormattingModelBuilder
import com.intellij.formatting.FormattingModelProvider
import raft.war.language.jass.JassLanguage.Companion.instance

internal class JassFormattingModelBuilder : FormattingModelBuilder {
    override fun createModel(formattingContext: FormattingContext): FormattingModel {
        val code = formattingContext.codeStyleSettings
        val jass = code.getCustomSettings(JassCodeStyleSettings::class.java)

        return FormattingModelProvider
            .createFormattingModelForPsiFile(
                formattingContext.containingFile,
                JassBlock(
                    node = formattingContext.node,
                    code = code,
                    common = code.getCommonSettings(instance.id),
                    jass = jass,
                ),
                code
            )
    }
}
